File: week14/15/3_Wed_NextGreater.py
from typing import List

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        res = [-1] * len(nums1)
        nums1_idx = {n:i for i, n in enumerate(nums1)}      # [4, 1, 2] => {4:0, 1:1, 2:2}

        for i in range(len(nums2)):
            if nums2[i] not in nums1_idx:
                continue
            for j in range(i+1, len(nums2)):
                if nums2[j] > nums2[i]:
                    idx = nums1_idx[nums2[i]]       # nums1_idx에서 nums2[i]의 인덱스(idx)를 가져옴 
                    res[idx] = nums2[j]
                    break
        return res

File: week14/15/3_Wed_NextGreater.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res(nums1.size(), -1);
        
        unordered_map<int, int> nums1_idx;
        for (int i=0; i < nums1.size(); i++) nums1_idx[nums1[i]] = i;

        for (int i=0; i < nums2.size(); i++) {
            if (nums1_idx.find(nums2[i]) == nums1_idx.end()) continue;          // vetor에서 해당값 찾기
            for (int j=i+1; j < nums2.size(); j++) {
                if (nums2[j] > nums2[i]) {
                    int idx = nums1_idx[nums2[i]];
                    res[idx] = nums2[j];
                    break;
                }
            }
        } 
        return res;
    }
};